optking/stepAlgorithms.py

```diff
@@ -1,825 +1,803 @@
+        dq_norm = sqrt(np.dot(dq, dq))
+        unit_dq = dq / dq_norm
+
+        # get gradient and hessian in step direction
+        grad = -1 * np.dot(fq, unit_dq)  # gradient, not force
+        hess = np.dot(unit_dq, np.dot(H, unit_dq))
+
+        self.logger.info("\t|target step| : %15.10f" % dq_norm)
+        self.logger.info("\tgradient     : %15.10f" % grad)
+        self.logger.info("\thessian      : %15.10f" % hess)
+
+        return dq_norm, unit_dq, grad, hess
+
+
+class OptimizationAlgorithm(OptimizationInterface):
+    """The standard minimization and transition state algorithms inherit from here. Defines
+    the take_step for those algorithms. Backstep and trust radius management are performed here.
+
+    All child classes implement a step() method using the forces and Hessian to compute
+    a step direction and possibly a step_length. trust_radius_on = False allows a child class
+    to override a basic trust radius enforcement."""
+
+    def __init__(self, molsys, history, params):
+        super().__init__(molsys, history, params)
+        self.trust_radius_on = True
+        self.intrafrag_trust = params.intrafrag_trust
+        self.intrafrag_trust_max = params.intrafrag_trust_max
+        self.intrafrag_trust_min = params.intrafrag_trust_min
+        self.consecutive_backsteps_allowed = params.consecutive_backsteps_allowed
+        self.dynamic_level = params.dynamic_level
+        self.opt_type = params.opt_type
+        self.linesearch = params.linesearch
+
+    @abstractmethod
+    def requires(self, **kwargs):
+        """Returns tuple with strings ('energy', 'gradient', 'hessian') for what the algorithm
+        needs to compute a new point"""
+        pass
+
+    def expected_energy(self, step, grad, hess):
+        """Compute the expected energy given the model for the step
+
+        Parameters
+        ----------
+        step: float
+            normalized step (unit length)
+        grad: np.ndarray
+            projection of gradient onto step
+        hess: np.ndarray
+            projection of hessian onto step
+        """
+        pass
+
+    @abstractmethod
+    def step(self, fq: np.ndarray, H: np.ndarray, *args, **kwargs) -> np.ndarray:
+        """Basic form of the algorithm"""
+        pass
+
+    def take_step(self, fq=None, H=None, energy=None, **kwargs):
+        """Compute step and take step"""
+
+        if len(fq) == 0:
+            self.logger.warning("Forces are missing. Step is 0")
+            return np.zeros(0)
+
+        self.history.append(self.molsys.geom, energy, fq)
+
+        if self.backstep_needed():
+            dq = self.backstep()
+        else:
+            dq = self.step(fq, H)
+
+        if self.trust_radius_on:
+            self.apply_intrafrag_step_scaling(dq)
+
+        self.molsys.interfrag_dq_discontinuity_correction(dq)
+        achieved_dq = displace_molsys(self.molsys, dq, fq)
+        dq_norm, unit_dq, projected_fq, projected_hess = self.step_metrics(dq, fq, H)
```

Comment:
For the projected hessian and energy, do you need to switch to achieved_dq?  Sometimes displace_molsys scales back step or just fails to take prescribed step.